### PR TITLE
04616 use ledgers for approval checks

### LINKED ISSUE #4616: `ApprovePrecompile` using base reality `StateView` for validation
### Description

When running [this EET](https://github.com/hashgraph/hedera-services/blob/04616-use-ledgers-for-approval-checks/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/hapi/ContractCallSuite.java#L1433) for [this contract](https://github.com/hashgraph/hedera-services/blob/04616-use-ledgers-for-approval-checks/test-clients/src/main/resource/contract/contracts/TestApprover/TestApprover.sol#L4) in branch `04616-use-ledgers-for-approval-checks`, we see `CONTRACT_REVERT_EXECUTED` and inconsistent child records,
```
REVERTED_SUCCESS
TOKEN_NOT_ASSOCIATED_TO_ACCOUNT
```
instead of a successful `ContractCreate`.

### Steps to reproduce

As above.

### Additional context

As in [closed issues](https://github.com/hashgraph/hedera-services/issues?q=is%3Aissue+is%3Aclosed+WorldLedgers), we have a problem because the "base reality" `StateView` is being used for validations inside `ApprovePrecompile` [here](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/precompile/impl/ApprovePrecompile.java#L208).

### PR DESCRIPTION
**Pull request by stoqnkpL**
**Description**:

**Related issue(s)**:

Fixes #4616 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/crypto/validators/DeleteAllowanceChecks.java [hunk 1]
```diff
@@ -64,6 +64,31 @@ public ResponseCodeEnum deleteAllowancesValidation(
...
             final Account payerAccount,
             final StateView view) {
+        final var accountStore = new AccountStore(validator, view.asReadOnlyAccountStore());
+        final var tokenStore =
+                new ReadOnlyTokenStore(
+                        accountStore,
+                        view.asReadOnlyTokenStore(),
+                        view.asReadOnlyNftStore(),
+                        view.asReadOnlyAssociationStore());
+        return deleteAllowancesValidation(nftAllowances, payerAccount, accountStore, tokenStore);
```
[the remark this answers is not in the document]
Comment (stoqnkpL) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/crypto/validators/DeleteAllowanceChecks.java:74]: This constructor is used only in the `CryptoDeleteAllowanceTransitionLogic` flow, I think it's safe to use `StateView` here (there are no updaters and stacked transactional ledgers in this case).

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/crypto/validators/ApproveAllowanceChecks.java [hunk 2]
```diff
@@ -74,6 +74,44 @@ public ResponseCodeEnum allowancesValidation(
...
+                        view.asReadOnlyNftStore(),
+                        view.asReadOnlyAssociationStore());
+
+        return allowancesValidation(
+                cryptoAllowances,
+                tokenAllowances,
+                nftAllowances,
+                payerAccount,
+                accountStore,
+                tokenStore);
```
[the remark this answers is not in the document]
Comment (stoqnkpL) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/crypto/validators/ApproveAllowanceChecks.java:93]: This constructor is used only in the `CryptoApproveAllowanceTransitionLogic` flow, I think it's safe to use `StateView` here (there are no updaters and stacked transactional ledgers in this case).

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/crypto/validators/DeleteAllowanceChecks.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (stoqnkpL) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/crypto/validators/DeleteAllowanceChecks.java:74]: The precompile flows build the stores using the transactional ledgers' current state and then call the other constructors of the `*Checks.java` classes, e.g.
```
public ResponseCodeEnum deleteAllowancesValidation(
            final List<NftRemoveAllowance> nftAllowances,
            final Account payerAccount,
            final AccountStore accountStore,
            final ReadOnlyTokenStore tokenStore) {
```

**Review by dimitar-dinev — APPROVED** [review]

**Review by IvanKavaldzhiev — APPROVED** [review]
LGTM

**Review by tinker-michaelj — APPROVED** [review]
LGTM

**Pull request merged**